File: shared/src/main/java/ru/privatenull/pnauth/platform/UnavailablePlatform.java
package ru.privatenull.pnauth.platform;

import ru.privatenull.pnauth.dialog.PlayerDialogs;
import ru.privatenull.pnauth.dialog.UnsupportedPlayerDialogs;
import java.time.Duration;
import java.util.Collection;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

/** Empty platform used until the hosting adapter has completed initialization. */
public final class UnavailablePlatform implements PnPlatform {
    private final PlayerDialogs dialogs = new UnsupportedPlayerDialogs();
    private final PlatformScheduler scheduler = new ImmediateScheduler();
    private final TaskRegistry tasks = new DefaultTaskRegistry(scheduler);

    @Override public PlatformType type() { return PlatformType.PAPER; }
    @Override public Optional<PnPlayer> player(UUID uniqueId) { return Optional.empty(); }
    @Override public Optional<PnPlayer> player(String username) { return Optional.empty(); }
    @Override public Collection<PnPlayer> players() { return List.of(); }
    @Override public PlatformScheduler scheduler() { return scheduler; }
    @Override public TaskRegistry tasks() { return tasks; }
    @Override public PlayerDialogs dialogs() { return dialogs; }

    private static final class ImmediateScheduler implements PlatformScheduler {
        @Override public TaskHandle execute(Runnable task) { task.run(); return DoneTask.INSTANCE; }
        @Override public TaskHandle execute(PnPlayer player, Runnable task) { return execute(task); }
        @Override public TaskHandle delayed(Duration delay, Runnable task) { return DoneTask.INSTANCE; }
        @Override public TaskHandle delayed(PnPlayer player, Duration delay, Runnable task) { return DoneTask.INSTANCE; }
        @Override public TaskHandle repeating(Duration initialDelay, Duration interval, Runnable task) { return DoneTask.INSTANCE; }
        @Override public TaskHandle repeating(PnPlayer player, Duration initialDelay, Duration interval, Runnable task) { return DoneTask.INSTANCE; }
    }

    private enum DoneTask implements TaskHandle {
        INSTANCE;
        @Override public boolean cancelled() { return true; }
        @Override public void cancel() { }
    }
}

File: shared/src/main/kotlin/ru/privatenull/pnauth/platform/UnavailablePlatform.kt
package ru.privatenull.pnauth.platform

import ru.privatenull.pnauth.dialog.PlayerDialogs
import ru.privatenull.pnauth.dialog.UnsupportedPlayerDialogs
import java.time.Duration
import java.util.Optional
import java.util.UUID

/** Empty platform used until the hosting adapter has completed initialization. */
class UnavailablePlatform : PnPlatform {
    private val dialogs: PlayerDialogs = UnsupportedPlayerDialogs()
    private val scheduler: PlatformScheduler = ImmediateScheduler()
    private val tasks: TaskRegistry = DefaultTaskRegistry(scheduler)

    override fun type(): PlatformType = PlatformType.PAPER
    override fun player(uniqueId: UUID): Optional<PnPlayer> = Optional.empty()
    override fun player(username: String): Optional<PnPlayer> = Optional.empty()
    override fun players(): Collection<PnPlayer> = emptyList<PnPlayer>()
    override fun scheduler(): PlatformScheduler = scheduler
    override fun tasks(): TaskRegistry = tasks
    override fun dialogs(): PlayerDialogs = dialogs

    private class ImmediateScheduler : PlatformScheduler {
        override fun execute(task: Runnable): TaskHandle {
            task.run()
            return DoneTask.INSTANCE
        }
        override fun execute(player: PnPlayer, task: Runnable): TaskHandle = execute(task)
        override fun delayed(delay: Duration, task: Runnable): TaskHandle = DoneTask.INSTANCE
        override fun delayed(player: PnPlayer, delay: Duration, task: Runnable): TaskHandle = DoneTask.INSTANCE
        override fun repeating(initialDelay: Duration, interval: Duration, task: Runnable): TaskHandle = DoneTask.INSTANCE
        override fun repeating(player: PnPlayer, initialDelay: Duration, interval: Duration, task: Runnable): TaskHandle = DoneTask.INSTANCE
    }

    private enum class DoneTask : TaskHandle {
        INSTANCE;
        override fun cancelled(): Boolean = true
        override fun cancel() {}
    }
}
